docs/RingBuffer.md

```diff
@@ -0,0 +1,303 @@
+ * @param[out] consumer pointer* to start of read-write mapped consumer page
+ *
+ * @returns Wait handle
+ */
+HANDLE ebpf_ring_buffer_get_wait_handle(const ebpf_map_t* map);
+```
+
+## Ringbuffer consumer
+
+### libbpf mapped memory consumer example
+
+
+This consumer directly accesses the records from the producer memory and directly updates the consumer offset to show the logic.
+
+```c++
+
+//
+// == Ringbuf helpers ==
+//
+
+// Ring buffer record is 64 bit header + data.
+typedef struct _rb_header
+{
+    //NOTE: bit fields are not portable, so this is just for simpler example code -- the actual code should use bit masking to perform equivalent operations on the header bits.
+    uint8_t locked : 1;
+    uint8_t discarded : 1;
+    uint32_t length : 30;
+    uint32_t offset; // for kernel use (offset of record in pages from start of buffer data area)
+} rb_header_t;
+
+typedef struct _rb_record
+{
+    rb_header_t header;
+    uint8_t data[];
+} rb_record_t;
+
+/**
+ * @brief clear the ringbuffer.
+ */
+void rb_flush(uint64_t *cons_offset, const uint64_t *prod_offset) {
+    *cons_offset = *prod_offset;
+}
+
+
+//
+// == mmap/epoll consumer ==
+//
+
+void *rb_cons; // Pointer to read/write mapped consumer page with consumer offset.
+void *rb_prod; // Pointer to start of read-only producer pages.
+
+// Open ringbuffer.
+fd_t map_fd = bpf_obj_get(rb_map_name.c_str());
+if (map_fd == ebpf_fd_invalid) return 1;
+
+auto rb = ring_buffer__new(
+    map_fd,
+    NULL, //callback function for callback-based consumer
+    nullptr, nullptr);
+if (!rb) return 1;
+
+// Initialize iocp wait handle.
+HANDLE wait_handle = ebpf_ring_buffer_get_wait_handle(rb);
+if (!wait_handle) {
+    // … log error …
+    goto Cleanup;
+}
+
+// get pointers to the producer/consumer pages
+int err = ebpf_ring_buffer_get_buffer(&rb_prod, &rb_cons);
+
+if (err) {
+    goto Cleanup;
+}
+
+const uint64_t *prod_offset = (const uint64_t*)rb_prod; // Producer offset ptr (r only).
+uint64_t *cons_offset = (uint64_t*)rb_cons; // Consumer offset ptr (r/w mapped).
+const uint8_t *rb_data = ((const uint8_t*)rb_prod) + PAGESIZE; // Double-mapped rb data ptr (r only).
+
+// have_data used to track whether we should wait for notification or just keep reading.
+bool have_data = *prod_offset > *cons_offset;
+
+void           *lp_ctx = NULL;
+OVERLAPPED     *overlapped = NULL;
+DWORD          bytesTransferred = 0;
+
+// Now loop until error.
+For(;;) {
+  if (!have_data) { // Only wait if we have already caught up.
+    // Wait for rb to notify -- or we could spin/poll on *prod_offset > *cons_offset.
+    DWORD wait_status = WaitForSingleObject(wait_handle, INFINITE);
+
+    if (wait_status != WAIT_OBJECT_0) { // No notification
+      uint32_t wait_err = GetLastError();
+      if (wait_err == /* terminal error */) {
+        // … log error …
+        break;
+      }
+      have_data = *prod_offset > *cons_offset; // It's possible we still have data.
```

Comment:
You need a ReadAcquire here (or an explicit memory barrier). With out that, you can potentially read the record before its been written to by the producer. Likewise you will need a WriteRelease in the producer.

Reply:
Also, reading it multiple times like this is going to cause issues. It should read prod_offset once using ReadAcquire and then use the value.

Reply:
Now uses ReadAcquire64/WriteRelease64 for access to the producer and consumer offset.

I went with bit fields for the record header for simpler example code, so that part is already pseudocode and I didn't add the ReadAcquire needed to read the header, but added a comment in the struct.